### ISSUE TITLE
Document/Improve: Continuous batching/chunking of a stream

### ISSUE THREAD

**Issue by jmealo** [brianc/node-pg-copy-streams#118]
I'm creating this issue to discuss solutions (documentation or otherwise) to help folks avoid some of the non-memory related pitfalls on #116 when continuously batching COPY commands.

See:
https://github.com/brianc/node-pg-copy-streams/issues/116#issuecomment-803643157

**Comment by jeromew** [comment]
Regarding the "antipattern" that is mentioned in https://github.com/brianc/node-pg-copy-streams/issues/118#issue-837158501

`pgStream.end() // calling this ends the COPY command`

This only starts the process of ending the COPY operation. You should not call

`pgStream = client.query(copyFrom(COPY_QUERY))`

and pipe into the rotated stream until the previous stream emits 'finish'.

In the code, what happens is :

```
COPY -> COMMITING
     -> COPY
```

when what you want in order to leave some breathing room to the backend is at least

```
COPY -> COMMITING -> COPY
```

**Comment by jeromew** [comment]
Just to make it clear this issue is for discussing a pattern that attracts some people because of the possibilities given by `pg-copy-stream`

Some people want to create a long-lived ingestion mechanism to ingest data into postgresql. For example, you could imagine that you have a never ending stream of logs that you want to ingest into postgresql.

this could be done by

1. inserting log lines one by one via INSERT commands
2. inserting batched volumes of log lines (let's say 100s) via MULTI-INSERT commands. cf "insert multiple rows" on https://www.postgresql.org/docs/14/dml-insert.html

there are probably other ways to do that with postgresql, including mechanisms linked with replication.

but still it can be interesting to see if and how COPY can help for this kind of scenarios.

first it is necessary to understand that a COPY FROM operation ingest rows. During the operation, the rows are not visible. They will become visible only once the COPY operation is terminated.

So `pg-copy-stream` cannot be used to open a long-lived stream, push data into it, and expect that this data will be queryable. For the data to be visible and queryable, you need to finish the COPY operation.

Once this is said, the idea of rotated COPY operations appear : ingest N rows via COPY and then create a new COPY operation.

It could be interesting to agree on advantages and disadvantages of the different solutions that can be used to do that and see how it impacts the performance of postgresql depending on the throughput of the ingested data, and how the throughput of out-of-band queries that need to be executed on the data.

I will leave this issue open for several months to see if people want to share their experience, have ideas around this, and discuss this further.

**Comment by jmealo** [comment]
@jeromew: Howdy, here I am bulk ingesting data into PostgreSQL again 😃  ... Multi-INSERT commands seems slow compared to copy but I haven't benchmarked it yet.

I'm realizing that a library that wraps `pg-copy-streams` that adds the logic for safely handling batches of bulk inserts, that flush for visibility and explains everything this issue does could be a good compromise.

If there was any interest in linking to it from this library, I'd gladly consider implementing it. Thoughts? @jeromew @brianc

**Comment by jeromew** [comment]
Hello,

I understand the need to find a nice streaming ingestion mechanism for long-lived streams and mega-huge ingested filed with regular visibility on the data beeing ingested (using one COPY operation only gives visibility after the COPY is commited). 

it would indeed be interesting to benchmark COPY vs Multi-Insert.

by Multi-Insert for N rows, I do not mean N INSERT operations, "1 insert per row", but I mean the ability that postgres has to insert several rows as a batch in one operation - cf https://www.postgresql.org/docs/current/dml-insert.html and thus send 1 INSERT with N rows

```
You can insert multiple rows in a single command:

INSERT INTO products (product_no, name, price) VALUES
    (1, 'Cheese', 9.99),
    (2, 'Bread', 1.99),
    (3, 'Milk', 2.99);
```

The TIP on this page says 

> When inserting a lot of data at the same time, consider using the [COPY](https://www.postgresql.org/docs/current/sql-copy.html) command. It is not as flexible as the [INSERT](https://www.postgresql.org/docs/current/sql-insert.html) command, but is more efficient.

but if the regular visibility is important to you I would be surprised that the time spent setting up regular COPY operation for N rows will outperform a regular INSERT with N rows. 

A specificity also when using COPY is that `copy-from` does not require the incoming stream to be respectful of row boundaries (Postgres has everything it needs to detect row boundaries and reconstruct rows from streamed network chunks) but in your solution you would need to either add a row boundary detection mechanism or require incoming streams to be respectful of boundaries.

Note also that if the stream is an object stream with 1 row per chunk, it is important performance-wise to recombine the rows into bigger network chunks as per https://github.com/brianc/node-pg-copy-streams/issues/127

**Comment by Inviz** [comment]
I'm seeing copy stream library spending a lot of time flushing stream synchronous. Is there any way to make it async?

**Comment by Inviz** [comment]
Thanks @jeromew . I am already batching the queries. I think my questions is more philosophical - is there a way to write into PG buffer without blocking JS world? Probably not? 

Then would spawning streamer in a worker with parallelism, - do you think stream-COPYing into the same table makes sense in parallel?

**Comment by jeromew** [comment]
I don't think `pg-copy-stream` is currently blocking JS world when writing into PG (well maybe it is but this would need a precise test script highlighting this). we're writing into a stream. There could be scenarios where if the upstream source itself is synchronous maybe it has precedence or something but here again, without a simple & precise test script hightlighing the issue there is not much i can do.

regarding multiple clients streaming into the same table, I haven't tested it. If postgres supports it, it could probably speed things up.
but according to [source code](https://github.com/postgres/postgres/blob/master/src/backend/commands/copy.c#L114C33-L114C49) it seems COPY FROM acquires a Row Exclusive Mode on the table so I don't think you can do 2 copy operations in parallel on the same table ; you will have to test.

**Comment by jeromew** [comment]
@Inviz I agree that the node stream machinery takes some toll on the main thread so having `pg-copy-streams` on a huge file in parallel with another application sharing the main thread can probably be seen ; but I think the main drivers on the total ingestion time are the network calls (which is probably why the batching stream in front of `from` improves things a lot)

why do you talk about 'worker' ? is it because you share the ingestion with another application like a web application ? For sure if you push stream pressure too far it can have an effect on the main thread. I think `pg-copy-streams` handle the back pressure correctly but never tested it thouroughly. maybe playing with the watermarks can help. Maybe there are other ways to give some breathing air to the main thread. Also not having postgres on the same machine will help if that's your case.

Can I ask what is your use case ? I would help me understand the scenario.

For now I am not in favor of going the 'native' way :
 - 1/ would the 'record' throughput really be implemented in javascript ? I now there are other implementations out there in other languages.
 - 2/ do you already handle your source stream as binary ? This can both shrink to total size to ingest & give less work to postgres ; but it will give more work on the javascript side if you need to convert from plain to binary
 - 3/ depending on the scenario and the data, there are other quick wins that can be implemented: suppress indices before ingestion ; ingest in many tables in parallel and then send a reconciliation sql request that merges those tables or have some kind of view on all the ingested tables, ..

**Comment by Inviz** [comment]
@jeromew I really appreciate your thorough reply, thank you. I hope that doesn't hijack the thread of discussion, as i think it's semi-related to the OP.

> why do you talk about 'worker' ? is it because you share the ingestion with another application like a web application ? For sure if you push stream pressure too far it can have an effect on the main thread. 

So as you very helpfully pointed out, COPY into a single table doesnt work in parallel. In my case however I am streaming into multiple tables at once (as a result of processing one input stream), so potentially putting different streams into different workers could increase the overall throughput, by making multiple streams not block each other when flushing chunks into PG. This seems to be the most promising way so far to unlock higher speeds. 

I think you are right that there could be alternative ways to give breathing space to the main thread, through balancing act of finding the right buffer size.   

Thanks for providing your comments on the native bindings approach. I think you are right, there's not a whole lot to be gained without making big changes, like going binary, suppressing indecies, streaming into temp tables, etc. I am using continuous aggregations/hypertables in timescaledb (PG extension), so it's desirable to find approach that keeps streaming realtime. But i think second best approach is streaming into temp table and then force PG to do the merging. 

I think in my workloads I am not at the point that makes it feasible. I wanted to figure out what's the best I could do without losing flexibility. 

I'll post up my results of experimenting with workers.

**Comment by jeromew** [comment]
@Inviz in the one-stream => multiple COPY stream to different tables scenario, and depending on the work, having the ability to use several CPUs via workers can probably increase the throughput.
I guess you have already found https://github.com/timescale/timescaledb-parallel-copy (go language) in which timescaledb probably put some thought. it seems to be using workers each responsible with batches of around ~ 5000 lines all inserting in the same hypertable. probably timescaledb has modified the COPY mechanism to benefit from hypertables so that could bypass the LOCK in some way if the batches fall into different physical tables).

Depending on the scenario you can also look into removing the wall logging on these tables. you could also maybe look into redis streams to have the initial PUSH into redis and then workers PULLing from redis streams but that's just an idea.

you can also take a look at [pg-copy-streams-binary](https://github.com/jeromew/node-pg-copy-streams-binary) . it is not optimized for high throuput because it does systematic byte<->object conversions but it can give you an example about binary handling and one streams => many tables. When I wrote that I used https://www.npmjs.com/package/multi-fork for the partitioning of the streams / this module has interesting ideas but I realize that it would probably need a refresh with newer versions of node / streams.

optimizing node pipelines is hard ! I am still looking for a node pipeline "live viewer" if you happen to find one !

**Comment by Inviz** [comment]
Yes I've seen that timescaledb parallel copy thing. When I dug deeper into pgx dependency, i didnt quite find where it achieves the concurrency while writing into a table, as it seems to be using COPY FROM faithfully all the way down. Timescaledb however is known for adding features that dont require turning them on explicitly, so it could be something inside, as you are theorizing.  The tool has some regressions in losing reporting capability, so i couldnt at the time make faithful test. I'll give it more thought, there's probably to be gained something here.

Thanks for your ideas and thanks for the pointer for binary version of the library and some context around it. I'll check it out! 

Personally i am working on my own async generator-based library https://github.com/Inviz/iterators which allows for easy concurrency, partitioning and such. But i definitely feel what you mean about live viewer. Understanding and tracing streams is hard. I had to use flamecharts mostly. But i was considering to build better tracing/viewing primitives.

**Comment by jeromew** [comment]
note you may also want to look at alternative js COPY implementations like in https://github.com/porsager/postgres
I did not test to see how they behave compared to pg/pg-copy-stream but maybe there is less overhead for your use case

**Comment by jeromew** [comment]
FWIW there is a blog post comparing different ways of copying a data table in postgres which could be interesting to you - https://ongres.com/blog/fastest_way_copy_data_between_postgres_tables/

I thought about one way that could be worth testing :

1 - COPY from node to postgres chunks of N rows in parallel to unlogged tables - cf https://www.crunchydata.com/blog/postgresl-unlogged-tables. For this you would need P clients and make sure that they indeed receive data in parallel from the source stream. The fastest I think would be to do it with a TEMPORARY TABLE with a sufficient temp buffers to receive enough data - cf https://dba.stackexchange.com/questions/285076/is-it-possible-to-have-a-temporary-on-the-fly-postgresql-table-which-only-ever
this first step would make sure the data arrives on the server.

2. Once the data in on the server (or at least once one of the TEMP table is considered full with its N rows) you can insert it into your main tables. you could also probably have a transaction around the COPY into TEMP + MOVE + DROP TEMP if needed for higher data protection. Note you could also "stream-prepare" the N rows chunks into files locally so it would be easier to replay a file for which the transaction failed (which will happen with a very high throughput)

I think this would give you a lot of throughput as the P clients would saturate the ingestion into postgres without beeing slowed by WAL, INDEX, VACCUM, ; the MOVE operation into the table/hypertable (which would happen fairly often and in parallel as the batches arrive)  is a then a postgres/timescaledb issue which probably has solutions to increase throughput if needed.

probably your can code that easily with `iterators` !

keep me posted on your results maybe this could result in a good answer to give to people that have this long-running ingestion use case.

**Comment by Inviz** [comment]
@jeromew Thank yo! I'd like to test some of these ideas, you are very kind to give me so many of them. I appreciate your input and your time, i dont take it as a granted. 

In my short term approach i went with batches of INSERTs, but this doesnt feel ideal to me. I do get 2x speed compared to blocking stream (which perhaps isnt optimized perfectly in my use case). However my opinion is that COPY _and_ streams together simply _must_ be better. Streaming into temp tables, and then doing a commit within PG - that sounds like a meaningful approach that is relatively easy to control. Coincidentally i did think about dynamic tuning of concurrency/batches to achieve the best throughput today - it's a fun idea.

I think in order of low-hangedness the fruits are:
1) Swap COPY implementation in hopes  that they have async write
2) COPY into temp table to see the performance profile of that
3) Attempt to parallelize node-pg-copy-stream by using worker - it's a bit annoying that I can't generate a worker on the fly, i would need to pass dependencies there, env variables, etc. Conceptually easy to do, just unfun
4) pg_bulkload according to that article seems to have great speed characteristics. I didn't yet look into what does it do to achieve that. I'd like to investigate

Doing things like disabling WAL, indexes, etc - seems like the last resort. I too am interested in finding a more or less generic approach, that is reproducable without risking of having db in weird state.

 I'll keep you posted, if you dont mind if it gets a bit asynchronous.

**Comment by jmealo** [comment]
Not sure if this helps you at all: https://www.timescale.com/blog/boosting-postgres-insert-performance


**TLDR:** This can deliver a 2x performance boost over `VALUES` in some cases:

```sql
INSERT INTO sensors (ts, sensorid, value) 
  SELECT * 
  FROM unnest(
    $1::timestamptz[], 
    $2::text[], 
    $3::float8[]
)
```

**Comment by jeromew** [comment]
@jmealo interesting hack thanks ! it gives even more leverage to the INSERT option for ingestion.

I did not do any performance test to see if it could be applied to the 2nd step in the pipeline (once the data in inside postgres), but this "unnest means query is prepared only once" performance hack can maybe be used.

the best I could come up with at this stage without a Pg/PSQL function is in a DO block

```
do
$$
declare
   agg_col1 integer[];
   agg_col2 text[];
begin
   -- select the number of actors from the actor table
   SELECT array_agg(col1), array_agg(col2) into agg_col1, agg_col2 from source limit 1;
   INSERT INTO target (col1, col2) select * from unnest(agg_col1, agg_col2);
end;
$$;
```

note this can also be used to create a table with

```
do
$$
declare
   agg_col1 integer[];
   agg_col2 text[];
begin
   -- select the number of actors from the actor table
   SELECT array_agg(col1), array_agg(col2) into agg_col1, agg_col2 from source limit 1;
   CREATE TABLE target as select * from unnest(agg_col1, agg_col2) as x(col1, col2);
end;
$$;
```

**Comment by jeromew** [comment]
@Inviz note also that there might be an "elephant in the room" here for some architectures.

using COPY FROM STDIN, you can only send postgres *uncompressed* data (it could be a great OPTION to add to a future version of postgres) so this is using more network than it should.

As I said earlier, you are probably forced to store interim files locally to manage failure modes.
If said files are compressed, why not send them to the postgres server disk via SCP and then COPY FROM locally after decompression.

All in all, with a patched postgres accepting compressed data on stdin, the following pipeline looks quite interesting :

source (local) => workers (local) => binary(worker) => compression(worker) => COPY FROM STDIN BINARY COMPRESSED (worker) => TEMP TABLE (postgres) => INSERT/UNNEST INTO target
along with a monitoring and dynamic optimization of the #workers / #clients / #size of batches / #latency

**Comment by jeromew** [comment]
@Inviz there was a link on HN in the last few days - https://pglocks.org/
according to this site, I/we could have misunderstood the effects of the locks applied to COPY.
this website seems to consider that 2 COPY FROM on the same table can happen concurrently.
the best would be to test it

**Comment by Inviz** [comment]
@jeromew super interesting, good find. Indeed it says COPY FROM can be concurrent with others. I'd like to give it a try.
